Introduce a fluent API to construct tensors from external data.

cbalioglu opened the pull request:
Summary:
This diff introduces the following changes and improvements:

- Introduces a new fluent API to construct tensors from external data as an alternative to `from_blob` overloads. See below for an example.
- Leverages several small-buffer optimizations which result in %50 reduction in tensor construction times.
- Exposes a new (lightweight) way to construct tensors by passing a naked `context` and `context_deleter` pair as an alternative to the existing `deleter` parameter.
- Updates the existing `from_blob` overloads to internally use the fluent API.

```
// Example 1
at::Tensor tensor = at::for_blob(data, sizes)
  .strides(strides)
  .context(context, [](void *ctx) { delete static_cast<Ctx*>(ctx); })
  .options(...)
  .target_device(...)
  .make_tensor();

// Example 2
at::Tensor tensor = at::for_blob(data, sizes).make_tensor();

// Example 3
at::Tensor tensor = at::for_blob(data, sizes)
  .deleter(...)
  .make_tensor();
```

Test Plan:
Below are the folly Benchmark results for the following two equivalent operations:

```
// The fluent API
at::Tensor tensor = at::for_blob(data, sizes)
  .deleter([buffer](void*) mutable { buffer.reset(); })
  .options(dtype(c10::ScalarType::Float))
  .make_tensor();

// The original `from_blob` overload
at::Tensor tensor = at::from_blob(
  data,
  sizes,
  [buffer](void*) mutable { buffer.reset(); },
  dtype(c10::ScalarType::Float));
```

```
============================================================================
scripts/balioglu/from_blob_exp/main.cpp         relative  time/iter  iters/s
============================================================================
fluent                                                     298.34ns    3.35M
from_blob                                         55.19%   540.51ns    1.85M
============================================================================
```

Various similar experiments show an approximate %50 reduction in tensor construction times.

Differential Revision: D27269344

ezyang commented:
> Leverages several small-buffer optimizations which result in %50 reduction in tensor construction times.

What are the small buffer optimizations? I skimmed through the PR and didn't see anything that jumped out.

cbalioglu commented:
> > Leverages several small-buffer optimizations which result in %50 reduction in tensor construction times.
> 
> What are the small buffer optimizations? I skimmed through the PR and didn't see anything that jumped out.

- Instead of `zero_sizes()` returning a `std::vector<int64_t>` `makeTempSizes()` returns a `SmallVector<int64_t, 5>`.
- We do not call `detail::defaultStrides()` anymore (which returns another heap-allocated vector that gets discarded right after constructing the tensor). If no strides are specified, we call `set_sizes_contiguous()` instead.
- The `deleter` requires a heap-allocated context deleter to be passed to `DataPtr`. With the `withContext()` we avoid that, by passing a context pointer and a raw function pointer.

ezyang commented on aten/src/ATen/templates/Functions.h:
```diff
@@ -64,19 +64,87 @@ TORCH_API at::Tensor conv3d(
+  TensorOptions opts_{};
+};
+
+inline TensorMaker forBlob(void* data, IntArrayRef sizes) noexcept {
```
camel case function name is inconsistent with other APIs (case in point: it's `at::from_blob`)

ezyang commented:
Thanks, the optimizations are much appreciated.

I think my primary commentary is about naming bikeshed: function name (`forBlob` versus `for_blob`) as well as naming of methods (`withStrides` versus `strides`). We have a preexisting set of helpers for defining fluent interfaces for NN modules, see torch/csrc/api/include/torch/arg.h and torch/csrc/api/include/torch/nn/options/linear.h; while you don't have to actually directly use the code here, it would be good if we lined up the conventions here.

cbalioglu commented:
> Thanks, the optimizations are much appreciated.

You are welcome!

> I think my primary commentary is about naming bikeshed: function name (`forBlob` versus `for_blob`) as well as naming of methods (`withStrides` versus `strides`). We have a preexisting set of helpers for defining fluent interfaces for NN modules, see torch/csrc/api/include/torch/arg.h and torch/csrc/api/include/torch/nn/options/linear.h; while you don't have to actually directly use the code here, it would be good if we lined up the conventions here.

Thanks a lot for the pointers. I am fairly new to the code base; good to know that we have an existing convention for fluent APIs. Let me check them. Hope to have a second revision later today.

cbalioglu committed:
Introduce a fluent API to construct tensors from external data. (#54530)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/54530

This diff introduces the following changes and improvements:

- Introduces a new fluent API to construct tensors from external data as an alternative to `from_blob` overloads. See below for an example.
- Leverages several small-buffer optimizations which result in %50 reduction in tensor construction times.
- Exposes a new (lightweight) way to construct tensors by passing a naked `context` and `context_deleter` pair as an alternative to the existing `deleter` parameter.
- Updates the existing `from_blob` overloads to internally use the fluent API.

```
// Example 1
at::Tensor tensor = at::for_blob(data, sizes)
  .strides(strides)
  .context(context, [](void *ctx) { delete static_cast<Ctx*>(ctx); })
  .options(...)
  .target_device(...)
  .make_tensor();

// Example 2
at::Tensor tensor = at::for_blob(data, sizes).make_tensor();

// Example 3
at::Tensor tensor = at::for_blob(data, sizes)
  .deleter(...)
  .make_tensor();
```

Test Plan:
Below are the folly Benchmark results for the following two equivalent operations:

```
// The fluent API
at::Tensor tensor = at::for_blob(data, sizes)
  .deleter([buffer](void*) mutable { buffer.reset(); })
  .options(dtype(c10::ScalarType::Float))
  .make_tensor();

// The original `from_blob` overload
at::Tensor tensor = at::from_blob(
  data,
  sizes,
  [buffer](void*) mutable { buffer.reset(); },
  dtype(c10::ScalarType::Float));
```

```
============================================================================
scripts/balioglu/from_blob_exp/main.cpp         relative  time/iter  iters/s
============================================================================
fluent                                                     298.34ns    3.35M
from_blob                                         55.19%   540.51ns    1.85M
============================================================================
```

Various similar experiments show an approximate %50 reduction in tensor construction times.

Differential Revision: D27269344

fbshipit-source-id: 26cad33986836ec6576bacb485c2a01c10ff8005

ezyang reviewed (approved):
yey

swolchok commented:
How does this new implementation ensure that the requested device, the device of the passed-in `data` `void` pointer, and the device specified in `TensorOptions` all match?

swolchok commented:
Also it seems unfortunate that for CPU we first make an empty CPU tensor with its own allocated `Storage` (see https://github.com/pytorch/pytorch/blob/2309173143abf3795e44a8472803d68849668434/aten/src/ATen/Utils.cpp#L66) and then throw that `Storage` away and provide our own. If we really did assume CPU tensors, we could avoid all that by using `detail::make_tensor` directly and providing our own `Storage` from the beginning.
Perhaps we could/should provide a dispatched way to make a `Tensor` with a given `Storage`? Not sure on that solution; there seems to be an extra degree of freedom in here somewhere that maybe we don't need...

cbalioglu commented:
Thanks for the feedback @swolchok. I deliberately did not touch the actual tensor construction logic. I just copied and cleaned up the majority of the existing implementation. Besides doing the optimizations I mentioned above and introducing a new API the internal machinery is more or less the same. As I am fairly new to the code base and this is such a core API, I did not want to accidentally break anything.

> How does this new implementation ensure that the requested device, the device of the passed-in data void pointer, and the device specified in TensorOptions all match?

I found the `target_device` parameter particularly confusing and still not sure about its use case; still, kept the existing logic intact. Similarly not sure why we are comparing the `target_device` and `TensorOptions`'s device only when a device index is specified.

As you already mentioned, I am fairly confident that there is room for improvement in how we construct tensors from external data. I would be happy to work further on this if deemed necessary.

ezyang commented:
Yes, an easy further improvement is to factor `empty_generic` further so that you can pass in a DataPtr directly instead of calling Allocate on allocator, then making `for_blob` use it.

cbalioglu commented:
> Yes, an easy further improvement is to factor `empty_generic` further so that you can pass in a DataPtr directly instead of calling Allocate on allocator, then making `for_blob` use it.

See #55705.
